### ISSUE TITLE
Allow custom_linear_solve to return things besides the solution

### ISSUE THREAD

**Issue by alewis** [google/jax#4322]
If I'm reading correctly `lax.custom_linear_solve` can presently only wrap `_solve` functions that return the computed solution `x` and nothing else. Particularly since in-place modification of arguments is of course not possible, this is a pretty severe limitation:

- Most solvers also return some kind of diagnostic information that the user will definitely want.
- Some solvers (e.g. LGMRES) additionally return auxiliary data that can be used to accelerate the solution of similar problems (the SciPy version of LGMRES does this with an in-place modification). This is quite useful in the fairly common case that the solve is one step in a larger iterative algorithm.

Is there a way around this?

**Comment by shoyer** [comment]
I agree, this does seem like an unfortunate limitation of the current implementation. My original hope was to solve this with solve general system for side effects in JAX code, but that doesn't seem to be forthcoming.

In the meantime, I think a good solution would be to add an optional `has_aux` argument to `custom_linear_solve`, which could work like `has_aux` on `jax.grad`: if `hax_aux` is True, then the user provided `solve` function (and maybe `transpose_solve`, too?) should return a tuple of the solution and any auxiliary output instead.

Note also that `jax.experimental.host_callback` is a good solution for the diagnostic information use case (especially for solves that appear in the backwards pass).
